cql3/selection/selection.cc

```diff
@@ -386,17 +388,61 @@ std::unique_ptr<result_set> result_set_builder::build() {
     return std::move(_result_set);
 }
 
-result_set_builder::restrictions_filter::restrictions_filter(::shared_ptr<restrictions::statement_restrictions> restrictions,
+namespace {
+
+using namespace expr;
+
+/// A visitor that copies an expression, but replaces token operator and unselected-column operators with the
+/// TRUE literal.
+struct drop_token_and_unselected_columns {
+    const selection& sel;
+
+    expression operator()(const conjunction& c) const {
+        auto r = c.children | std::views::transform([this] (const expression& e) { return std::visit(*this, e); });
+        return conjunction{std::vector<expression>(r.begin(), r.end())};
+    }
+
+    expression operator()(const binary_operator& binop) const {
+        return seastar::visit(binop.lhs,
+                [] (const token&) { return expression(true); },
+                [&] (const column_value& cv) {
+                    return sel.has_column(*cv.col) ? expression(binop) : expression(true);
+                },
+                [&] (auto&& default_case) { return expression(binop); });
+    }
+
+    expression operator()(bool b) const { return b; }
+};
+
+/// Compiles an expression equivalent to the given restrictions.  A filter can evaluate this expression on each
+/// row to satisfy the restrictions.
+expression compile(
+        const restrictions::statement_restrictions& restrictions,
+        const query_options& options,
```

Comment:
If we get query_options here, how is this compilation? It's performed on every query.

Reply:
In fact you don't use options, which suggests the caller could be during prepare (doesn't have to be in this patch).

Reply:
> If we get query_options here, how is this compilation? It's performed on every query.

That doesn't make it not-compilation. :) There are a few phases in every query's life:
1. preparation
2. execution
 2.1. compute bounds
 2.2. fetch rows
 2.3. create restrictions_filter
 2.4. for each fetched row, apply the restrictions_filter

The code above is in phase 2.3, which happens on every query but not every row.  By compiling (or optimizing, I don't care what we call it) the query here, we avoid repetitive tasks in 2.4.  And we need query_options to turn markers into concrete values (eg, so we can compile regexes for LIKE patterns).

> In fact you don't use options

I'll take `options` out of this patch, then add it right back in the next one, which replaces LIKE nodes with regex match.

Reply:
Ok. I think we'll want to move this to the prepare phase eventually, and have an extra, optional step for expressions that have non-literal LIKE patterns. We don't want to penalize all other expressions for that.

How far are we from generating expressions directly from the parser?

Reply:
> How far are we from generating expressions directly from the parser?

I was going to work on that next, after this patch and [translating LIKE to compiled regex](https://github.com/dekimir/scylla/commits/match_regex-in-ast2).  So I'm guessing start in a couple of days, and finish... IDK when, depends on landmines I uncover along the way. ;)